File: lc/coreprep/implementation/practice_iterations/patterns_iterations_1/number_range.cpp
#include <algorithm>
#include <iostream>
#include <list>
#include <queue>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

enum Pos { Min, Max };

int find_pos(const vector<int> &a, int low, int high, int k, Pos pos) {
  int last_pos = -1;
  while (low <= high) {
    int mid = low + (high - low) / 2;

    if (a[mid] < k) {
      low = mid + 1;
    } else if (a[mid] > k) {
      high = mid - 1;
    } else {
      last_pos = mid;

      if (pos == Min) {
        high = mid - 1;
      } else {
        low = mid + 1;
      }
    }
  }

  return last_pos;
}

// O(logn) / O(1)
pair<int, int> find_range(const vector<int> &a, int k) {
  int min_pos = -1, max_pos = -1;

  min_pos = find_pos(a, 0, a.size() - 1, k, Pos::Min);
  max_pos = find_pos(a, 0, a.size() - 1, k, Pos::Max);

  return make_pair<>(min_pos, max_pos);
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, k;
  cin >> tt;

  while (tt--) {
    cin >> n >> k;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    auto [pos1, pos2] = find_range(a, k);
    cout << pos1 << " " << pos2 << "\n";
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_2/NumberRange.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class NumberRange {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int k = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        var res = findMinMaxPos(a, 0, n - 1, k);
        System.out.printf("%d %d\n", res[0], res[1]);
      }
    }
  }

  static enum Find {
    MIN,
    MAX;
  }

  // O(logn) / O(1)
  public static int[] findMinMaxPos(int[] a, int low, int high, int k) {
    int minPos = binarySearch(a, low, high, k, Find.MIN);
    int maxPos = binarySearch(a, low, high, k, Find.MAX);

    return new int[] {minPos, maxPos};
  }

  private static int binarySearch(int[] a, int low, int high, int k,
                                  Find find) {
    int lastPos = -1;
    while (low <= high) {
      int mid = low + (high - low) / 2;

      if (a[mid] < k) {
        low = mid + 1;
      } else if (a[mid] > k) {
        high = mid - 1;
      } else {
        lastPos = mid;

        if (find == Find.MIN) {
          high = mid - 1;
        } else {
          low = mid + 1;
        }
      }
    }

    return lastPos;
  }
}